executor/new_executor.go

```diff
@@ -149,12 +223,86 @@ func (e *HashJoinExec) prepare() error {
 		}
 	}
 
+	e.resultRows = make(chan *Row, e.concurrency*1000)
+	e.resultErr = make(chan error, 1)
+
+	e.wg = sync.WaitGroup{}
+	for i := 0; i < e.concurrency; i++ {
+		e.wg.Add(1)
+		go e.doJoin(i)
+	}
+	go e.closeChanWorker()
+
 	e.prepared = true
 	return nil
 }
 
-func (e *HashJoinExec) constructMatchedRows(bigRow *Row) (matchedRows []*Row, err error) {
-	hasNull, hashcode, err := getHashKey(e.bigHashKey, bigRow, e.targetTypes)
+func (e *HashJoinExec) closeChanWorker() {
+	e.wg.Wait()
+	close(e.resultRows)
+}
+
+// do join job
+func (e *HashJoinExec) doJoin(idx int) {
+	for {
+		var (
+			bigRows []*Row
+			ok      bool
+			err     error
+		)
+		select {
+		case bigRows, ok = <-e.bigTableRows[idx]:
+		case err = <-e.bigTableErr:
+		}
+		if err != nil {
+			e.resultErr <- errors.Trace(err)
+			break
+		}
+		if !ok {
+			e.wg.Done()
+			break
+		}
+		for _, bigRow := range bigRows {
+			succ := e.join(idx, bigRow)
+			if !succ {
+				break
+			}
+		}
+	}
+}
+
+func (e *HashJoinExec) join(idx int, bigRow *Row) bool {
+	var (
+		matchedRows []*Row
+		err         error
+	)
+	bigMatched := true
+	if e.bigFilter != nil {
+		bigMatched, err = expression.EvalBoolInShard(e.bigFilter, bigRow.Data, e.ctx, idx)
```

Comment:
There is another way to keep concurrent safe.
1. Create a struct named execContext which contains: e.bigTableRows[idx], deep cloned e.bigFilter, e.hashKeyBuffer[idx]
2. Each doJoin goroutine get an execContext instance as parameter.